### PR TITLE
Fix pvalue behaviour

### PR DESCRIPTION
**Pull request by snath-xoc**
This PR addS shuffle indices and is_clusterer method as well as updates reports. Attempts to fix p-value behaviour under np.random.choice as identified by @jeremiedbb in issue #14 

TO DO:

- [ ] The test seems to still be unstable: some times 13 estimators pass, other times 17 estimators.
- [x] DecisionTreeClassifier was only sensitive when min_samples_split was changed otherwise predicitons are the same: this does not solve the p-value=1 problem for AdaBoostClassifier
- [ ] ExtraTreeClassifier (which has DecisionTreeClassifier at the back) is still having the problem where p-values are 1

### CODE REVIEW

**Comment by jeremiedbb** [comment]
> shuffle indices

I think it's missing this part :)

**Comment by snath-xoc** [comment]
Odd, I thought I had added it in but may have messed up when I updated it with the main branch. It should be in now... all estimators fail now though (if you look at the report). I tried checking for silly mistakes but cannot see any (it is the same fix as I put in PR #13 ), maybe some fresh eyes will help?

**Comment by snath-xoc** [comment]
@ogrisel and @jeremiedbb the p-value fix for issue #14 is implemented here properly now. The test still is a bit sensitive (see TO DO), perhaps incorporating what we had discussed during the meeting (i.e., running it several times and plotting the p-values). What do you think?

File: src/sample_weight_audit/sklearn_stochastic_params.py [hunk 1]
```diff
@@ -32,25 +32,27 @@
...
     },
+    BisectingKMeans: {"n_clusters": 10},
     LinearSVR: {"dual": True},
     LinearSVC: {"dual": True},
     Ridge: {"solver": "sag"},
     Lasso: {"selection": "random"},
     LassoCV: {"selection": "random"},
     ElasticNet: {"selection": "random"},
     ElasticNetCV: {"selection": "random"},
+    DecisionTreeClassifier: {"max_features": 0.5, "min_samples_split": 0.3},
```
Comment (ogrisel) [inline src/sample_weight_audit/sklearn_stochastic_params.py:48]: We should not use non-default values for `min_samples_split` or `min_samples_leaf` in our sample weight equivalence checks. Those would be expected to fail to be sample_weight/repetition equivalent. We might actually want to document that in their docstring (by cross-linking to the glossary entry for sample weights).

Instead, let's use `min_weight_fraction_leaf=0.2` or similar in `src/sample_weight_audit/sklearn_stochastic_params.py`.

Comment (snath-xoc) [reply src/sample_weight_audit/sklearn_stochastic_params.py:48]: Yes agreed, still fishy that we get p-values of exactly 1, am trying to find the root of it, but it may be just what we get as a result... but 1 every time is fishy.

Comment (ogrisel) [reply src/sample_weight_audit/sklearn_stochastic_params.py:48]: I agree, and this is not the only one. The equivalence tests for the MLP models and other estimators still have a p-value of 1 consistently.

**Comment by ogrisel** [comment]
> The test seems to still be unstable: some times 13 estimators pass, other times 17 estimators.

This is expected if we do not pass a fixed `random_state` value to `check_weighted_repeated_estimator_fit_equivalence`.

File: src/sample_weight_audit/estimator_check.py [hunk 2]
```diff
@@ -179,7 +179,7 @@ def compute_predictions(est, X):
 
 
 def check_pipeline_and_fit(est, X, y, sample_weight=None, seed=None):
-    if not is_classifier(est) and not is_regressor(est) and hasattr(est, "transform"):
+    if not is_classifier(est) and not is_regressor(est) and not is_clusterer(est):
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:182]: I think this change was not meant to be included in this PR. Let me revert.

File: src/sample_weight_audit/estimator_check.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 from dataclasses import dataclass
 
 import numpy as np
-from sklearn.base import clone, is_classifier, is_regressor
+from sklearn.base import clone, is_classifier, is_regressor, is_clusterer
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:4]:
```suggestion
from sklearn.base import clone, is_classifier, is_regressor
```

File: src/sample_weight_audit/estimator_check.py (hunk 2 above)
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:182]:
```suggestion
    if not is_classifier(est) and not is_regressor(est) and hasattr(est, "transform"):
```

**Review by ogrisel — APPROVED** [review]

**Comment by ogrisel** [comment]
I also implemented a Bonferroni correction scheme + increased the number of fits so that actual bugs are more likely to be detected.

**Comment by ogrisel** [comment]
@snath-xoc feel free to merge this PR into `main` if you agree with the solution and then merge `main` into #13 to resume work on the score-based method.

In parallel, it would be great to start opening PRs to investigate and fix the worst offenders. I suspect that bagging based models should be fixed by using the weights for resampling rather than routing to the base estimators.

**Comment by ogrisel** [comment]
Also note: `HistGradientBoostingClassifier/Regressor` have a `min_samples_leaf=20` by default. We might want to set that to `min_samples_leaf=1` and open an issue to have `min_weight_fraction_leaf` hyper-parameter in the future instead.

**Comment by snath-xoc** [comment]
Thank you for the help @ogrisel , glad to see that there are no estimators with p-value of 1 now so at least it wasn't a problem with the estimators' configurations themselves.
